modules/tensor_mechanics/src/postprocessors/MaterialTensorAverage.C

```diff
@@ -0,0 +1,71 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MaterialTensorAverage.h"
+
+registerMooseObject("TensorMechanicsApp", MaterialTensorAverage);
+registerMooseObject("TensorMechanicsApp", ADMaterialTensorAverage);
+
+template <bool is_ad>
+InputParameters
+MaterialTensorAverageTempl<is_ad>::validParams()
+{
+  InputParameters params = MaterialTensorIntegralTempl<is_ad>::validParams();
+  params.addClassDescription("Computes the average of a material property over a volume.");
+  return params;
+}
+
+template <bool is_ad>
+MaterialTensorAverageTempl<is_ad>::MaterialTensorAverageTempl(const InputParameters & parameters)
+  : MaterialTensorIntegralTempl<is_ad>(parameters), _volume(0.0)
+{
+}
+
+template <bool is_ad>
+void
+MaterialTensorAverageTempl<is_ad>::initialize()
+{
+  MaterialTensorIntegralTempl<is_ad>::initialize();
+
+  _volume = 0.0;
+}
+
+template <bool is_ad>
+void
+MaterialTensorAverageTempl<is_ad>::execute()
+{
+  MaterialTensorIntegralTempl<is_ad>::execute();
+
+  _volume += this->_current_elem_volume;
+}
+
+template <bool is_ad>
+Real
+MaterialTensorAverageTempl<is_ad>::getValue()
+{
+  const Real integral = MaterialTensorIntegralTempl<is_ad>::getValue();
+
+  this->gatherSum(_volume);
+
+  return integral / _volume;
+}
+
+template <bool is_ad>
+void
+MaterialTensorAverageTempl<is_ad>::threadJoin(const UserObject & y)
+{
+  MaterialTensorAverageTempl<is_ad>::threadJoin(y);
+
+  const MaterialTensorAverageTempl<is_ad> & pps =
+      static_cast<const MaterialTensorAverageTempl<is_ad> &>(y);
+  _volume += pps._volume;
+}
```

Comment:
I have no idea how to correct this given the recursivity error (it has always been obscure how `threadjoin` works).
@lindsayad any advice?

Reply:
You probably mean to call the base class thread join method right?
```c++
MaterialTensorIntegralTempl<is_ad>::threadJoin(y)
```

Reply:
yes... what miserable mistake